.github/workflows/ci.yml

```diff
@@ -1,19 +1,33 @@
-name: build
+name: Build Status
 on:
+  workflow_dispatch:
   push:
+    branches: [master]
   pull_request:
-    types: [opened, reopened]
+    branches: [master]
 
 jobs:
-  test:
+  build:
     runs-on: ubuntu-latest
+    env:
+      CI: true
+
     strategy:
       matrix:
-        version: [8, lts/*, latest]
+        node-version: [8.x, 10.x, 12.x, 14.x, 16.x, 18.x]
+
     steps:
-      - uses: actions/checkout@HEAD
-      - uses: actions/setup-node@HEAD
+      - uses: actions/checkout@v3
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v3
         with:
-          node-version: ${{ matrix.version }}
-      - run: npm install
+          node-version: ${{ matrix.node-version }}
+      - run: npm ci
       - run: npm test
+      - run: npm update
+      - run: npm ci
+      - run: npm test
```

Comment:
Couple of things:
- Why the need to run `npm ci; npm test` twice? are we trying to test the codebase against the deps defined in package-lock as well as the deps after updating said lock?
- You may consider the use of `if` to [conditionally](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) run commands, so that for newer node versions we run `npm ci`, and for older ones we run `npm install`.
  - This is of course dependent on whether we want to continue supporting node 4 and 6. Limiting ourselves to node 8 would give us the freedom to rewrite the codebase with more modern language constructs (like async/await and `const` and `let` instead of `var`), but that depends on whether we get round to doing so.

Reply:
I'm acting as if we are going to do a `semver-major` release to drop older versions of Node. I really don't see a need to keep supporting something that old.


> are we trying to test the codebase against the deps defined in package-lock as well as the deps after updating said lock?

Yes, that is the idea. We had some problems with `node-saml` and `passport-saml` where people would report bugs that we couldn't reproduce. We found it was actually due to an incompatibility with a dependency. We actually found a problem with a TypeScript incompatibility/regression. Maybe I can mark these as `experimental` so that we don't fail the entire PR on them? I just don't want problem like that to remain hidden for very long, so I hesitate to do even that.